src/drivers/virtual_timer_controller.cpp

```diff
@@ -0,0 +1,136 @@
+#include "drivers/virtual_timer_controller.hpp"
+
+#include "drivers/virtual_timer_linked_list.hpp"
+#include "nrf52833.h"
+#include "scheduler/pending_process_callbacks.hpp"
+
+namespace edge::drivers {
+
+extern "C" {
+void TIMER3_IRQHandler(void)
+{
+    VirtualTimerController vtc = VirtualTimerController::get();
+    vtc.TIMER->EVENTS_COMPARE[2] = 0;
+    vtc.update_ll();
+    return;
+}
+}
+
+VirtualTimerController::VirtualTimerController()
+{
+    TIMER->BITMODE = 0x3;
+    TIMER->MODE = 0x0;
+    TIMER->PRESCALER = 0x4;
+    TIMER->TASKS_START = 1;
+    TIMER->INTENSET = 0x40000;
+    NVIC_EnableIRQ(TIMER3_IRQn);
+}
+
+void VirtualTimerController::update_ll()
+{
+    __disable_irq();
+    node_t* ptr = list_get_first();
+    if (!ptr) {
+        return;
+    }
+    uint32_t val = ptr->timer_value;
+    uint32_t current_time = read_timer();
+    while (val < current_time) {
+        list_remove_first();
+        PendingProcessCallbacks::get().add_ready_callback(
+            0, reinterpret_cast<void (*)(int, int)>(ptr->callback)
+        );
+        free(ptr);
+
+        ptr = list_get_first();
+        if (!ptr) {
+            break;
+        }
+        val = ptr->timer_value;
+        current_time = read_timer();
+    }
+    __enable_irq();
+    return;
```

Comment:
why are we disabling interrupts for this

Reply:
We did this in lab2 bruh